fix: sui-select in plunkr starter template not working

zone.js:645 Error: Uncaught (in promise): TypeError: deepExtend_2 is not a function
TypeError: deepExtend_2 is not a function
    at deepExtend (ng2-semantic-ui.umd.js:226)
    at SuiLocalizationService.exports.SuiLocalizationService.SuiLocalizationService.get (ng2-semantic-ui.umd.js:258)
    at SuiSelectBase.onLocaleUpdate (ng2-semantic-ui.umd.js:13255)
    at SuiSelectBase (ng2-semantic-ui.umd.js:13038)
    at new SuiSelect (ng2-semantic-ui.umd.js:13828)
    at createClass (provider.ts:365)
    at createDirectiveInstance (provider.ts:182)
    at createViewNodes (view.ts:341)
    at callViewAction (view.ts:782)
    at execComponentViewsAction (view.ts:700)
    at deepExtend (
...

http://plnkr.co/edit/rYbB9u25oJCzHWA2GbqA?p=preview

--- a/demo/src/app/pages/demo-pages.module.ts
+++ b/demo/src/app/pages/demo-pages.module.ts
@@ -1,6 +1,6 @@
 import { NgModule } from "@angular/core";
 import { CommonModule } from "@angular/common";
-import { FormsModule } from "@angular/forms";
+import { FormsModule, ReactiveFormsModule } from "@angular/forms";
 import { RouterModule } from "@angular/router";
 import { SuiModule } from "../../../../src";
 import { DemoComponentsModule } from "../components/demo-components.module";
@@ -38,6 +38,7 @@ import { TestPage } from "./development/test/test.page";
     imports: [
         CommonModule,
         FormsModule,
+        ReactiveFormsModule,
         RouterModule,
         SuiModule,
         DemoComponentsModule

--- a/demo/src/app/pages/development/test/test.page.html
+++ b/demo/src/app/pages/development/test/test.page.html
@@ -7,16 +7,6 @@
 <demo-page-content>
     <h2 class="ui dividing header">Examples</h2>
     <div class="ui segment">
-        <div class="ui fluid left icon input">
-            <i class="calendar icon"></i>
-            <input suiDatepicker
-                   [(ngModel)]="date"
-                   pickerMode="month"
-                   [pickerMaxDate]="maxDate"
-                   [pickerMinDate]="minDate"
-                   [pickerFirstDayOfWeek]="3">
-        </div>
-        <br>
-        <p>{{ date }}</p>
+        
     </div>
 </demo-page-content>

--- a/demo/src/app/pages/development/test/test.page.ts
+++ b/demo/src/app/pages/development/test/test.page.ts
@@ -1,26 +1,10 @@
 import { Component, AfterViewInit, ViewChild, TemplateRef } from "@angular/core";
+import { FormControl, Validators } from "@angular/forms";
 
 @Component({
     selector: "demo-page-test",
     templateUrl: "./test.page.html"
 })
 export class TestPage {
-    public date:Date = new Date();
-
-    public minDate:Date;
-    public maxDate:Date;
-
-    constructor() {
-        const today = new Date();
-
-        this.minDate = new Date(today.getFullYear(), today.getMonth() - 2);
-        this.maxDate = new Date(today.getFullYear(), today.getMonth() + 2);
-
-        console.log(this.minDate);
-        console.log(this.maxDate);
-    }
-
-    public log(item:any):void {
-        console.log(item);
-    }
+    constructor() {}
 }

--- a/scripts/bundle.js
+++ b/scripts/bundle.js
@@ -14,15 +14,14 @@ const defaultPlugins = [
     }),
     commonjs({
         namedExports: {
-            "node_modules/bowser/src/bowser.js": ["mobile"],
-            "node_modules/deep-extend/lib/deep-extend.js": ["call"]
+            "node_modules/bowser/src/bowser.js": ["mobile"]
         },
         include: [
             "node_modules/element-closest/**",
             "node_modules/popper.js/**",
             "node_modules/date-fns/**",
             "node_modules/bowser/src/bowser.js",
-            "node_modules/deep-extend/lib/deep-extend.js"
+            "node_modules/deep-extend/**"
         ]
     })
 ]

--- a/src/behaviors/localization/services/localization.service.ts
+++ b/src/behaviors/localization/services/localization.service.ts
@@ -1,15 +1,13 @@
 import { Injectable, EventEmitter } from "@angular/core";
-import { ILocaleValues, IPartialLocaleValues } from "../interfaces/values";
+import { ILocaleValues, IPartialLocaleValues, RecursivePartial } from "../interfaces/values";
 import enGB from "../locales/en-GB";
-import * as extend from "deep-extend";
-import { RecursivePartial } from "../interfaces/partial";
+
+// Oh rollup.
+import * as $deepExtend from "deep-extend";
+const deepExtend:<T, U>(target:T, source:U) => T & U = ($deepExtend as any).default || $deepExtend;
 
 function deepClone<T>(obj:T):T {
     return JSON.parse(JSON.stringify(obj));
-}
-
-function deepExtend<T, U>(target:T, source:U):T & U {
-    return extend.call(undefined, target, source);
 }
 
 function lang(language:string):string {

--- a/src/modules/accordion/components/accordion-panel.ts
+++ b/src/modules/accordion/components/accordion-panel.ts
@@ -1,4 +1,4 @@
-import { Component, Input, Output, EventEmitter } from "@angular/core";
+import { Component, Input, Output, EventEmitter, ChangeDetectorRef } from "@angular/core";
 import { SuiAccordionService } from "../services/accordion.service";
 import { TransitionController, Transition } from "../../transition";
 
@@ -36,6 +36,7 @@ export class SuiAccordionPanel {
 
     public set service(service:SuiAccordionService) {
         this._service = service;
+        this._changeDetector.detectChanges();
     }
 
     @Input()
@@ -88,7 +89,7 @@ export class SuiAccordionPanel {
     @Output()
     public isOpenChange:EventEmitter<boolean>;
 
-    constructor() {
+    constructor(private _changeDetector:ChangeDetectorRef) {
         this.transitionController = new TransitionController(false);
 
         this._isOpen = false;

--- a/src/modules/accordion/components/accordion.ts
+++ b/src/modules/accordion/components/accordion.ts
@@ -59,7 +59,7 @@ export class SuiAccordion implements AfterContentInit {
         this.updatePanels();
 
         // Reconnect panels after they have updated.
-        this._panels.changes.subscribe(() => setTimeout(() => this.updatePanels()));
+        this._panels.changes.subscribe(() => this.updatePanels());
     }
 
     public updatePanels():void {

--- a/src/modules/checkbox/components/checkbox.ts
+++ b/src/modules/checkbox/components/checkbox.ts
@@ -31,8 +31,11 @@ export class SuiCheckbox implements ICustomValueAccessorHost<boolean> {
     @HostBinding("class.checked")
     public isChecked:boolean;
 
-    @Output()
-    public checkChange:EventEmitter<boolean>;
+    @Output("checkChange")
+    public onCheckChange:EventEmitter<boolean>;
+
+    @Output("touched")
+    public onTouched:EventEmitter<void>;
 
     @Input()
     public isDisabled:boolean;
@@ -54,12 +57,18 @@ export class SuiCheckbox implements ICustomValueAccessorHost<boolean> {
 
     constructor() {
         this.isChecked = false;
-        this.checkChange = new EventEmitter<boolean>();
+        this.onCheckChange = new EventEmitter<boolean>();
+        this.onTouched = new EventEmitter<void>();
 
         this.isDisabled = false;
         this.isReadonly = false;
 
         this._checkboxClasses = true;
+    }
+
+    @HostListener("mousedown", ["$event"])
+    public onMouseDown(e:MouseEvent):void {
+        e.preventDefault();
     }
 
     @HostListener("click")
@@ -70,9 +79,14 @@ export class SuiCheckbox implements ICustomValueAccessorHost<boolean> {
         }
     }
 
+    @HostListener("focusout")
+    public onFocusOut():void {
+        this.onTouched.emit();
+    }
+
     public toggle():void {
         this.isChecked = !this.isChecked;
-        this.checkChange.emit(this.isChecked);
+        this.onCheckChange.emit(this.isChecked);
     }
 
     public writeValue(value:boolean):void {
@@ -86,7 +100,10 @@ export class SuiCheckbox implements ICustomValueAccessorHost<boolean> {
 
 @Directive({
     selector: "sui-checkbox",
-    host: { "(checkChange)": "onChange($event)" },
+    host: {
+        "(checkChange)": "onChange($event)",
+        "(touched)": "onTouched()"
+    },
     providers: [customValueAccessorFactory(SuiCheckboxValueAccessor)]
 })
 export class SuiCheckboxValueAccessor extends CustomValueAccessor<boolean, SuiCheckbox> {

--- a/src/modules/checkbox/components/radiobutton.ts
+++ b/src/modules/checkbox/components/radiobutton.ts
@@ -38,8 +38,11 @@ export class SuiRadioButton<T> implements ICustomValueAccessorHost<T> {
 
     public currentValue:T;
 
-    @Output()
-    public currentValueChange:EventEmitter<T>;
+    @Output("currentValueChange")
+    public onCurrentValueChange:EventEmitter<T>;
+
+    @Output("touched")
+    public onTouched:EventEmitter<void>;
 
     @Input()
     public isDisabled:boolean;
@@ -61,7 +64,8 @@ export class SuiRadioButton<T> implements ICustomValueAccessorHost<T> {
 
     constructor() {
         this.isChecked = false;
-        this.currentValueChange = new EventEmitter<T>();
+        this.onCurrentValueChange = new EventEmitter<T>();
+        this.onTouched = new EventEmitter<void>();
 
         this.isDisabled = false;
         this.isReadonly = false;
@@ -69,14 +73,24 @@ export class SuiRadioButton<T> implements ICustomValueAccessorHost<T> {
         this._radioClasses = true;
     }
 
+    @HostListener("mousedown", ["$event"])
+    public onMouseDown(e:MouseEvent):void {
+        e.preventDefault();
+    }
+
     @HostListener("click")
     public onClick():void {
         if (!this.isDisabled && !this.isReadonly) {
             this.currentValue = this.value;
-            this.currentValueChange.emit(this.currentValue);
+            this.onCurrentValueChange.emit(this.currentValue);
             this.update();
             this.focusRadio();
         }
+    }
+
+    @HostListener("focusout")
+    public onFocusOut():void {
+        this.onTouched.emit();
     }
 
     public update():void {
@@ -95,7 +109,10 @@ export class SuiRadioButton<T> implements ICustomValueAccessorHost<T> {
 
 @Directive({
     selector: "sui-radio-button",
-    host: { "(currentValueChange)": "onChange($event)" },
+    host: {
+        "(currentValueChange)": "onChange($event)",
+        "(touched)": "onTouched()"
+    },
     providers: [customValueAccessorFactory(SuiRadioButtonValueAccessor)]
 })
 export class SuiRadioButtonValueAccessor<T> extends CustomValueAccessor<T, SuiRadioButton<T>> {

--- a/src/modules/datepicker/directives/input.directive.ts
+++ b/src/modules/datepicker/directives/input.directive.ts
@@ -3,7 +3,7 @@ import { Directive, Host, Input, ElementRef, HostBinding, HostListener } from "@
 import { DateUtil, DatePrecision } from "../../../misc/util";
 import { SuiLocalizationService } from "../../../behaviors/localization";
 import { PopupTrigger } from "../../popup";
-import { SuiDatepickerDirective } from "./datepicker.directive";
+import { SuiDatepickerDirective, SuiDatepickerDirectiveValueAccessor } from "./datepicker.directive";
 import { InternalDateParser, DateParser } from "../classes/date-parser";
 import * as bowser from "bowser";
 
@@ -84,6 +84,7 @@ export class SuiDatepickerInputDirective {
     }
 
     constructor(@Host() public datepicker:SuiDatepickerDirective,
+                @Host() public valueAccessor:SuiDatepickerDirectiveValueAccessor,
                 public element:ElementRef,
                 localizationService:SuiLocalizationService) {
         this.useNativeOnMobile = true;
@@ -122,5 +123,10 @@ export class SuiDatepickerInputDirective {
             return this.datepicker.writeValue(parsed);
         }
         return this.datepicker.writeValue(undefined);
+    }
+
+    @HostListener("focusout")
+    public onFocusOut():void {
+        this.valueAccessor.onTouched();
     }
 }

--- a/src/modules/dropdown/directives/dropdown.ts
+++ b/src/modules/dropdown/directives/dropdown.ts
@@ -118,11 +118,9 @@ export class SuiDropdown implements AfterContentInit {
 
     @HostListener("focusout", ["$event"])
     private onFocusOut(e:FocusEvent):void {
-        setTimeout(() => {
-            if (!this._element.nativeElement.contains(document.activeElement)) {
-                this.externallyClose();
-            }
-        });
+        if (!this._element.nativeElement.contains(e.relatedTarget)) {
+            this.externallyClose();
+        }
     }
 
     @HostListener("keypress", ["$event"])

--- a/src/modules/search/components/search.ts
+++ b/src/modules/search/components/search.ts
@@ -242,11 +242,9 @@ export class SuiSearch<T> implements AfterViewInit {
 
     @HostListener("focusout", ["$event"])
     private onFocusOut(e:FocusEvent):void {
-        setTimeout(() => {
-            if (!this._element.nativeElement.contains(document.activeElement)) {
-                this.dropdownService.setOpenState(false);
-            }
-        });
+        if (!this._element.nativeElement.contains(e.relatedTarget)) {
+            this.dropdownService.setOpenState(false);
+        }
     }
 
     @HostListener("document:click", ["$event"])

--- a/src/modules/select/classes/select-base.ts
+++ b/src/modules/select/classes/select-base.ts
@@ -366,12 +366,10 @@ export abstract class SuiSelectBase<T, U> implements AfterContentInit {
 
     @HostListener("focusout", ["$event"])
     private onFocusOut(e:FocusEvent):void {
-        setTimeout(() => {
-            if (!this._element.nativeElement.contains(document.activeElement)) {
-                this.dropdownService.setOpenState(false);
-                this.onTouched.emit();
-            }
-        });
+        if (!this._element.nativeElement.contains(e.relatedTarget)) {
+            this.dropdownService.setOpenState(false);
+            this.onTouched.emit();
+        }
     }
 
     @HostListener("keypress", ["$event"])

--- a/src/modules/select/classes/select-base.ts
+++ b/src/modules/select/classes/select-base.ts
@@ -1,6 +1,6 @@
 import {
     ViewChild, HostBinding, ElementRef, HostListener, Input, ContentChildren, QueryList,
-    AfterContentInit, TemplateRef, ViewContainerRef, ContentChild
+    AfterContentInit, TemplateRef, ViewContainerRef, ContentChild, EventEmitter, Output
 } from "@angular/core";
 import { Subscription } from "rxjs/Subscription";
 import { DropdownService, SuiDropdownMenu } from "../../dropdown";
@@ -222,6 +222,9 @@ export abstract class SuiSelectBase<T, U> implements AfterContentInit {
     @Input()
     public transitionDuration:number;
 
+    @Output("touched")
+    public onTouched:EventEmitter<void>;
+
     constructor(private _element:ElementRef, protected _localizationService:SuiLocalizationService) {
         this.dropdownService = new DropdownService();
         // We do want an empty query to return all results.
@@ -236,6 +239,8 @@ export abstract class SuiSelectBase<T, U> implements AfterContentInit {
         this.icon = "dropdown";
         this.transition = "slide down";
         this.transitionDuration = 200;
+
+        this.onTouched = new EventEmitter<void>();
 
         this._selectClasses = true;
     }
@@ -364,6 +369,7 @@ export abstract class SuiSelectBase<T, U> implements AfterContentInit {
         setTimeout(() => {
             if (!this._element.nativeElement.contains(document.activeElement)) {
                 this.dropdownService.setOpenState(false);
+                this.onTouched.emit();
             }
         });
     }

--- a/src/modules/select/classes/select-base.ts
+++ b/src/modules/select/classes/select-base.ts
@@ -169,7 +169,13 @@ export abstract class SuiSelectBase<T, U> implements AfterContentInit {
 
     public get labelGetter():(obj:T) => string {
         // Helper function to retrieve the label from an item.
-        return (obj:T) => Util.Object.readValue<T, string>(obj, this.labelField);
+        return (obj:T) => {
+            const label = Util.Object.readValue<T, string>(obj, this.labelField);
+            if (label) {
+                return label.toString();
+            }
+            return "";
+        };
     }
 
     @Input()
@@ -187,11 +193,11 @@ export abstract class SuiSelectBase<T, U> implements AfterContentInit {
 
     public get configuredFormatter():(option:T) => string {
         if (this._optionFormatter) {
-            return r => this._optionFormatter!(r, this.isSearchable ? this.query : undefined);
+            return o => this._optionFormatter!(o, this.isSearchable ? this.query : undefined);
         } else if (this.searchService.optionsLookup) {
-            return r => this.labelGetter(r);
+            return o => this.labelGetter(o);
         } else {
-            return r => this.searchService.highlightMatches(this.labelGetter(r), this.query || "");
+            return o => this.searchService.highlightMatches(this.labelGetter(o), this.query || "");
         }
     }
 

--- a/src/modules/select/components/multi-select.ts
+++ b/src/modules/select/components/multi-select.ts
@@ -175,7 +175,10 @@ export class SuiMultiSelect<T, U> extends SuiSelectBase<T, U> implements ICustom
 // Value accessor directive for the select to support ngModel.
 @Directive({
     selector: "sui-multi-select",
-    host: { "(selectedOptionsChange)": "onChange($event)" },
+    host: {
+        "(selectedOptionsChange)": "onChange($event)",
+        "(touched)": "onTouched()"
+    },
     providers: [customValueAccessorFactory(SuiMultiSelectValueAccessor)]
 })
 export class SuiMultiSelectValueAccessor<T, U> extends CustomValueAccessor<U[], SuiMultiSelect<T, U>> {

--- a/src/modules/select/components/select.ts
+++ b/src/modules/select/components/select.ts
@@ -141,7 +141,10 @@ export class SuiSelect<T, U> extends SuiSelectBase<T, U> implements ICustomValue
 // Value accessor directive for the select to support ngModel.
 @Directive({
     selector: "sui-select",
-    host: { "(selectedOptionChange)": "onChange($event)" },
+    host: {
+        "(selectedOptionChange)": "onChange($event)",
+        "(touched)": "onTouched()"
+    },
     providers: [customValueAccessorFactory(SuiSelectValueAccessor)]
 })
 export class SuiSelectValueAccessor<T, U> extends CustomValueAccessor<U, SuiSelect<T, U>> {

